File: tools/ghidra_compare_metrics.py
from ghidra.program.model.symbol import SourceType
from ghidra.program.model.listing import CodeUnit


def count_comments(program):
    listing = program.getListing()
    it = listing.getCodeUnits(True)
    total = 0
    by_type = {
        "pre": 0,
        "post": 0,
        "eol": 0,
        "plate": 0,
        "repeatable": 0,
    }
    while it.hasNext() and not monitor.isCancelled():
        cu = it.next()
        has_any = False
        pre = cu.getComment(CodeUnit.PRE_COMMENT)
        if pre:
            by_type["pre"] += 1
            has_any = True
        post = cu.getComment(CodeUnit.POST_COMMENT)
        if post:
            by_type["post"] += 1
            has_any = True
        eol = cu.getComment(CodeUnit.EOL_COMMENT)
        if eol:
            by_type["eol"] += 1
            has_any = True
        plate = cu.getComment(CodeUnit.PLATE_COMMENT)
        if plate:
            by_type["plate"] += 1
            has_any = True
        repeatable = cu.getComment(CodeUnit.REPEATABLE_COMMENT)
        if repeatable:
            by_type["repeatable"] += 1
            has_any = True
        if has_any:
            total += 1
    return total, by_type


def main():
    program = currentProgram
    fm = program.getFunctionManager()
    st = program.getSymbolTable()

    all_symbol_count = 0
    user_defined_symbol_count = 0
    imported_symbol_count = 0
    analysis_symbol_count = 0
    default_symbol_count = 0

    user_defined_function_count = 0
    sample_user_functions = []

    sym_it = st.getAllSymbols(True)
    while sym_it.hasNext() and not monitor.isCancelled():
        sym = sym_it.next()
        all_symbol_count += 1
        source = sym.getSource()
        if source == SourceType.USER_DEFINED:
            user_defined_symbol_count += 1
        elif source == SourceType.IMPORTED:
            imported_symbol_count += 1
        elif source == SourceType.ANALYSIS:
            analysis_symbol_count += 1
        elif source == SourceType.DEFAULT:
            default_symbol_count += 1

    fn_it = fm.getFunctions(True)
    function_count = 0
    while fn_it.hasNext() and not monitor.isCancelled():
        fn = fn_it.next()
        function_count += 1
        if fn.getSymbol().getSource() == SourceType.USER_DEFINED:
            user_defined_function_count += 1
            if len(sample_user_functions) < 40:
                sample_user_functions.append(
                    "%s @ %s" % (fn.getName(), fn.getEntryPoint())
                )

    info = program.getOptions("Program Information")
    info_names = list(info.getOptionNames())
    interesting_info = {}
    for key in sorted(info_names):
        if key in [
            "Executable Format",
            "Executable Location",
            "Executable MD5",
            "Executable SHA256",
            "Compiler",
            "Compiler ID",
            "Language ID",
            "PE Property[TimeDateStamp]",
            "PE Property[OriginalFilename]",
            "PE Property[FileVersion]",
            "PE Property[ProductVersion]",
            "Created With Ghidra Version",
        ]:
            interesting_info[key] = info.getValueAsString(key)

    total_comment_units, comment_breakdown = count_comments(program)

    print("PROGRAM_NAME=%s" % program.getName())
    print("IMAGE_BASE=%s" % program.getImageBase())
    for key in sorted(interesting_info.keys()):
        print("INFO_%s=%s" % (key.replace(" ", "_"), interesting_info[key]))
    print("FUNCTION_COUNT=%d" % function_count)
    print("USER_DEFINED_FUNCTION_COUNT=%d" % user_defined_function_count)
    print("SYMBOL_COUNT=%d" % all_symbol_count)
    print("USER_DEFINED_SYMBOL_COUNT=%d" % user_defined_symbol_count)
    print("IMPORTED_SYMBOL_COUNT=%d" % imported_symbol_count)
    print("ANALYSIS_SYMBOL_COUNT=%d" % analysis_symbol_count)
    print("DEFAULT_SYMBOL_COUNT=%d" % default_symbol_count)
    print("COMMENT_UNITS=%d" % total_comment_units)
    for key in ["plate", "pre", "eol", "repeatable", "post"]:
        print("COMMENT_%s=%d" % (key.upper(), comment_breakdown[key]))
    for sample in sample_user_functions:
        print("USER_FUNCTION=%s" % sample)


main()

File: tools/GhidraCompareMetrics.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.List;

import ghidra.app.script.GhidraScript;
import ghidra.program.model.listing.CodeUnit;
import ghidra.program.model.listing.Function;
import ghidra.program.model.listing.FunctionManager;
import ghidra.program.model.listing.Listing;
import ghidra.program.model.listing.Program;
import ghidra.program.model.symbol.SourceType;
import ghidra.program.model.symbol.Symbol;
import ghidra.program.model.symbol.SymbolIterator;
import ghidra.program.model.symbol.SymbolTable;
import ghidra.framework.options.Options;

public class GhidraCompareMetrics extends GhidraScript {

	private static final class CommentCounts {
		int any;
		int plate;
		int pre;
		int eol;
		int repeatable;
		int post;
	}

	@Override
	protected void run() throws Exception {
		Program program = currentProgram;
		FunctionManager functionManager = program.getFunctionManager();
		SymbolTable symbolTable = program.getSymbolTable();

		int symbolCount = 0;
		int userDefinedSymbolCount = 0;
		int importedSymbolCount = 0;
		int analysisSymbolCount = 0;
		int defaultSymbolCount = 0;

		SymbolIterator symbolIterator = symbolTable.getAllSymbols(true);
		while (symbolIterator.hasNext() && !monitor.isCancelled()) {
			Symbol symbol = symbolIterator.next();
			symbolCount++;
			SourceType source = symbol.getSource();
			if (source == SourceType.USER_DEFINED) {
				userDefinedSymbolCount++;
			}
			else if (source == SourceType.IMPORTED) {
				importedSymbolCount++;
			}
			else if (source == SourceType.ANALYSIS) {
				analysisSymbolCount++;
			}
			else if (source == SourceType.DEFAULT) {
				defaultSymbolCount++;
			}
		}

		int functionCount = 0;
		int userDefinedFunctionCount = 0;
		List<String> sampleUserFunctions = new ArrayList<>();

		for (Function function : functionManager.getFunctions(true)) {
			if (monitor.isCancelled()) {
				break;
			}
			functionCount++;
			if (function.getSymbol().getSource() == SourceType.USER_DEFINED) {
				userDefinedFunctionCount++;
				if (sampleUserFunctions.size() < 40) {
					sampleUserFunctions.add(function.getName() + " @ " + function.getEntryPoint());
				}
			}
		}

		CommentCounts commentCounts = countComments(program);
		Options info = program.getOptions("Program Information");
		HashSet<String> interestingNames = new HashSet<>(Arrays.asList(
			"Executable Format",
			"Executable Location",
			"Executable MD5",
			"Executable SHA256",
			"Compiler",
			"Compiler ID",
			"Language ID",
			"PE Property[TimeDateStamp]",
			"PE Property[OriginalFilename]",
			"PE Property[FileVersion]",
			"PE Property[ProductVersion]",
			"Created With Ghidra Version"
		));

		println("PROGRAM_NAME=" + program.getName());
		println("IMAGE_BASE=" + program.getImageBase());
		for (String optionName : info.getOptionNames()) {
			if (interestingNames.contains(optionName)) {
				Object value = info.getObject(optionName, null);
				println("INFO_" + optionName.replace(' ', '_') + "=" + value);
			}
		}
		println("FUNCTION_COUNT=" + functionCount);
		println("USER_DEFINED_FUNCTION_COUNT=" + userDefinedFunctionCount);
		println("SYMBOL_COUNT=" + symbolCount);
		println("USER_DEFINED_SYMBOL_COUNT=" + userDefinedSymbolCount);
		println("IMPORTED_SYMBOL_COUNT=" + importedSymbolCount);
		println("ANALYSIS_SYMBOL_COUNT=" + analysisSymbolCount);
		println("DEFAULT_SYMBOL_COUNT=" + defaultSymbolCount);
		println("COMMENT_UNITS=" + commentCounts.any);
		println("COMMENT_PLATE=" + commentCounts.plate);
		println("COMMENT_PRE=" + commentCounts.pre);
		println("COMMENT_EOL=" + commentCounts.eol);
		println("COMMENT_REPEATABLE=" + commentCounts.repeatable);
		println("COMMENT_POST=" + commentCounts.post);
		for (String sample : sampleUserFunctions) {
			println("USER_FUNCTION=" + sample);
		}
	}

	private CommentCounts countComments(Program program) {
		CommentCounts counts = new CommentCounts();
		Listing listing = program.getListing();
		for (CodeUnit codeUnit : listing.getCodeUnits(true)) {
			if (monitor.isCancelled()) {
				break;
			}
			boolean hasAny = false;
			if (codeUnit.getComment(CodeUnit.PLATE_COMMENT) != null) {
				counts.plate++;
				hasAny = true;
			}
			if (codeUnit.getComment(CodeUnit.PRE_COMMENT) != null) {
				counts.pre++;
				hasAny = true;
			}
			if (codeUnit.getComment(CodeUnit.EOL_COMMENT) != null) {
				counts.eol++;
				hasAny = true;
			}
			if (codeUnit.getComment(CodeUnit.REPEATABLE_COMMENT) != null) {
				counts.repeatable++;
				hasAny = true;
			}
			if (codeUnit.getComment(CodeUnit.POST_COMMENT) != null) {
				counts.post++;
				hasAny = true;
			}
			if (hasAny) {
				counts.any++;
			}
		}
		return counts;
	}
}
